sql/tests: TestRandomSyntaxSelect failed

sql/tests.TestRandomSyntaxSelect [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/21312950?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/21312950?buildTab=artifacts#/) on release-26.1 @ [6b3864d7da4bf057d291eeb6ee1367a4bab9c195](https://github.com/cockroachdb/cockroach/commits/6b3864d7da4bf057d291eeb6ee1367a4bab9c195):


```
    rsg_test.go:125: Last executed (4): SELECT NOT ROW ( ) OVERLAPS ( ) IS NOT NAN TYPES FROM INDEX_AFTER_ORDER_BY_BEFORE_AT . BIGINT @ { NO_ZIGZAG_JOIN , FORCE_INDEX = OUT , FORCE_INVERTED_INDEX } WHERE 0 IS NOT OF ( DATE ARRAY [ 195 ] )
    rsg_test.go:125: Last executed (5): SELECT * , DEFAULT CONFLICT FROM ident 
    rsg_test.go:125: Last executed (6): SELECT UNIQUE ( error ILIKE ROW ( ) OVERLAPS ROW ( ) AS FLOAT FROM ident 
    rsg_test.go:125: Last executed (7): SELECT * , UNIQUE ( error AS THEN , * FROM ( ( error FETCH FIRST ROW ONLY FOR KEY SHARE OF INDEX_BEFORE_PAREN FOR NO KEY UPDATE SKIP LOCKED ) AS OF SYSTEM TIME UNIQUE ( error WHERE UNIQUE ( error TYPECAST GEOMETRY ( LINESTRINGM ) ARRAY [ 152 ]
    rsg_test.go:125: Last executed (8): SELECT ROW ( ) OVERLAPS ( ) IS NOT TRUE CHECK FROM ident 
    rsg_test.go:125: Last executed (9): SELECT UNIQUE ( error AS BUCKET_COUNT , * FROM ident 
    rsg_test.go:125: Last executed (10): SELECT DEFAULT IS NOT TRUE AS EXISTS FROM ( [ SHOW USERS error ] AS LEAST FULL OUTER LOOKUP JOIN [ DELETE error ] USING ( POINT , ERRORS ) ) WHERE UNIQUE ( error TYPEANNOTATE OUTER . ident . FROM ARRAY [ -300 ]
    rsg_test.go:125: Last executed (11): SELECT * , EXISTS ( ( error ORDER BY PRIMARY KEY FAMILY DESC ) FROM ident 
    rsg_test.go:125: Last executed (12): SELECT @ -436 FROM [ -92 ( -250 ) AS TRIGGER ] WITH ORDINALITY , LATERAL QUERIES . COLUMN . FAMILY ( error WITH ORDINALITY AS OF SYSTEM TIME DEFAULT WHERE SQRT NOT UNIQUE ( error LESS_EQUALS DEFAULT
    rsg_test.go:125: Last executed (13): SELECT DEFAULT COLLATE WHERE CLOSE FROM ( [ -193 ( ) DETACHED ] @ [ -129 ] ident ( REGION ) JOIN LATERAL ROWS FROM ( CURRENT_CATALOG ident ( SKIP ident ARRAY [ 31 ] , PLANS INTERVAL ARRAY [ 254 ] [ error , ident @ 25 [ -225 ] ) ) USING ( TIMETZ , SURVIVE ) ) WITH ORDINALITY AS ident ( GEOMETRY ) WHERE FAMILY . FAMILY . GREATEST . *
    rsg_test.go:125: Last executed (14): SELECT B'10010' NONVOTERS FROM ( ( VALUES error ORDER BY PRIMARY KEY MULTILINESTRINGZ . CHARACTERISTICS ASC , INDEX_AFTER_ORDER_BY_BEFORE_AT FAMILY @ COALESCE DESC ) ) CONVERSION WHERE @ 84
    rsg_test.go:125: Last executed (15): SELECT UNIQUE ( error IS NOT FALSE SHOW FROM [ -135 ( ) AS ident ] WITH ORDINALITY AS CHARACTERISTICS WHERE OPERATOR ( ROUTINES . CONTAINS ) DEFAULT IS NOT OF ( POINT error [ 240 ] , NONE . DELIMITER [ ] )
    rsg_test.go:125: Last executed (16): SELECT * , DEFAULT LEFT , * , * FROM [ 304 ( -313 , 113 , 435 ) SECURITY ( BIGINT , INTERVAL , ident ) ] @ { FAMILY = [ 90 ] } WITH ORDINALITY SEARCH , ( UNSPLIT ident JOIN [ EXPLAIN error ] USING ( LC_CTYPE , CHANGEFEED , DETAILS ) ) WHERE + OPERATOR ( FIRST_CONTAINS ) NOT DEFAULT TYPEANNOTATE @ -39 ARRAY [ 22 ] [ error
    rsg_test.go:125: Last executed (17): SELECT * , NOT UNIQUE ( error BACKWARD , * FROM [ UPDATE error ] , ONLY ( INDEX_BEFORE_PAREN . CHAR ) @ AUTHORIZATION WITH ORDINALITY WHERE OPERATOR ( # ) DEFAULT IS NOT TRUE IS FALSE IS NULL
    rsg_test.go:125: Last executed (18): SELECT ROW ( ) OVERLAPS ROW ( ) AS NOCREATEROLE FROM [ 308 STOP ] @ [ -89 ] AS ident ( ident , ident ) WHERE DEFAULT NOT SIMILAR TO NOT UNIQUE ( error IS OF ( ident . ARRAY . FAMILY ARRAY , VECTOR ARRAY [ -69 ] [ error )
    rsg_test.go:125: Last executed (19): SELECT ~ EXISTS ( ( error FOR NO KEY UPDATE NOWAIT ) FROM LATERAL SESSION_USER ( ) WHERE CBRT VECTOR . ident . TIME . *
    rsg_test.go:125: Last executed (20): SELECT UNIQUE ( error IS OF ( POLYGON error ARRAY [ 54 ] ) AS CHARACTER FROM ident 
    rsg_test.go:125: Last executed (21): SELECT UNIQUE ( error AS PARENT FROM [ -215 ( ) AS VALID ( CONVERT ) ] WHERE DEFAULT IS OF ( POLYGON error ARRAY [ -361 ] , @ 37 [ ] , ISNULL ARRAY [ -101 ] [ error )
    rsg_test.go:125: Last executed (22): SELECT DEFAULT CONTAINS UNIQUE ( error COLLATE POSITION AS ident FROM LATERAL [ FUNCTION 201 ] ( error AS BETWEEN WHERE ( ) OVERLAPS ROW ( ) TYPEANNOTATE ident ARRAY
    rsg_test.go:125: Last executed (23): SELECT * , OPERATOR ( DISTANCE ) DEFAULT ident FROM ( WITH VARBIT ( INOUT ) AS MATERIALIZED ( INSPECT TABLE BIGINT WITH OPTIONS DETACHED , DETACHED ) ( error INTERSECT ALL ( error FOR UPDATE OF REFRESH . ident , EXISTS . WINDOW . NULLIF NOWAIT FOR NO KEY UPDATE SKIP LOCKED ) WHERE NOT - 13 IS OF ( POLYGON error ARRAY [ 61 ] , INTERVAL SECOND ARRAY [ 127 ] [ error , @ 246 ARRAY , POLYGON error ARRAY [ 354 ] ) IS NOT UNKNOWN
    rsg_test.go:125: Last executed (24): SELECT * , DEFAULT SPLIT FROM ident 
    rsg_test.go:125: Last executed (25): SELECT ROW ( ) OVERLAPS ( ) AS PRIMARY FROM ident 
    rsg_test.go:125: Last executed (26): SELECT ROW ( ) OVERLAPS ( ) AS ASYMMETRIC FROM [ 47 ( -495 ) AS DECIMAL ( ident , ROW ) ] @ ident AS ident WHERE ~ TRUE IS NOT NULL
    rsg_test.go:125: Last executed (27): SELECT DEFAULT TYPEANNOTATE @ -55 ARRAY FROM ident 
    rsg_test.go:125: Last executed (28): SELECT NULL AS EXPERIMENTAL_FINGERPRINTS , * FROM FAMILY . ident WITH ORDINALITY , [ -333 ( 206 , -367 , 428 ) AS DROP ( VIRTUAL_CLUSTER_NAME , BIGINT ) ] @ [ 194 ] WITH ORDINALITY UNSAFE_RESTORE_INCOMPATIBLE_VERSION , ( ( TABLE error ORDER BY PRIMARY KEY FAMILY . COMPLETIONS DESC ) ) AS ADMIN ( VIRTUAL_CLUSTER_NAME , ident ) , ( ( [ -278 PRIVILEGES ( ident ) ] @ [ -405 ] WITH ORDINALITY JOIN LATERAL CURRENT_TIMESTAMP WITH ORDINALITY AS DOUBLE ( ident , ident , NAN ) USING ( NOCANCELQUERY , COVERING , ident , ident ) ) ) WHERE CHAR VARYING ( 243 ) 'string'
    rsg_test.go:125: Last executed (29): SELECT * , UNIQUE ( error POLICY FROM [ 486 ( -91 ) AS READ ] @ { FORCE_INDEX = [ -375 ] , NO_INDEX_JOIN } WITH ORDINALITY AS CONFIGURE WHERE FALSE + DEFAULT
    rsg_test.go:125: Last executed (30): SELECT ROW ( ) OVERLAPS ROW ( ) AS SESSIONS FROM LATERAL ( ( ( WITH RECURSIVE ident AS MATERIALIZED ( TRUNCATE TABLE ONLY ( FAMILY ) RESTRICT ) , LEVEL ( ident , ident ) AS NOT MATERIALIZED ( RESTORE INDEX_AFTER_ORDER_BY_BEFORE_AT . VARIADIC . WITHOUT FROM error ) , EXTRACT_DURATION ( IFERROR ) AS MATERIALIZED ( SHOW SESSION TIME ZONE ) TABLE [ -349 ( 290 ) AS POLYGON ] @ { IGNORE_FOREIGN_KEYS , DESC , FORCE_ZIGZAG = [ 31 ] , FORCE_ZIGZAG = WORK } WITH ORDINALITY IFERROR FOR READ ONLY ) OFFSET - -7.277028547980946e+43 ROW FOR UPDATE OF INDEX_BEFORE_NAME_THEN_PAREN SKIP LOCKED ) ) WITH ORDINALITY NATURAL INNER MERGE JOIN ONLY INDEX_AFTER_ORDER_BY_BEFORE_AT . TO CROSS HASH JOIN ( ( [ SHOW CLUSTER STATEMENTS ] WITH ORDINALITY OUT ( ident , DECIMAL , OVERLAY , DOMAIN , ident ) JOIN [ -348 ( -163 , -358 ) AS VISIBILITY ] @ [ 250 ] WITH ORDINALITY USING ( SETTINGS ) ) ) JOIN ONLY ( INDEX_AFTER_ORDER_BY_BEFORE_AT ) WITH ORDINALITY AS ident JOIN [ 356 ( ) TIMESTAMPTZ ( VIRTUAL_CLUSTER_NAME , GEOGRAPHY , ROWS ) ] @ TIMESTAMP WITH ORDINALITY AS COMMIT ( ident , ident ) USING ( IF , FUNCTIONS ) USING ( ident , CHARACTERISTICS , FILTER ) WHERE NOT DEFAULT ISNULL IS NAN
    rsg_test.go:125: Last executed (31): SELECT UNIQUE ( error NOT_REGIMATCH ROW ( ) OVERLAPS ( ) IS FALSE FROM [ SHOW INDEX FROM DATABASE ident ] WITH ORDINALITY WHERE OPERATOR ( COS_DISTANCE ) ROW ( ) OVERLAPS ROW ( )
    rsg_test.go:125: Last executed (32): SELECT ~ UNIQUE ( error AS ACTION FROM ident 
    rsg_test.go:125: Last executed (33): SELECT * , UNIQUE ( error TRIGGERS FROM INDEX_BEFORE_NAME_THEN_PAREN . ident * @ EXCLUDING AS ident WHERE FALSE COLLATE DEFERRABLE
    rsg_test.go:125: Last executed (34): SELECT * , NOT DEFAULT IN ( ) FROM ident 
    rsg_test.go:125: Last executed (35): SELECT NOT EXISTS ( ( WITH IF ( ident ) AS ( DROP ROLE error ) ( error ) ) COLLATE ANY IS NAN ident FROM ident 
    rsg_test.go:125: Last executed (36): SELECT ( WITH LOCAL ( ident , TIMESTAMPTZ , SEQUENCES ) AS MATERIALIZED ( IMPORT INTO INDEX_BEFORE_NAME_THEN_PAREN ident DATA ( ident ) ) TABLE error FOR SHARE OF DEC , FAMILY . FAMILY , AVAILABILITY , WRITE . TRIM . SEQUENCES , FAMILY , ident . ESCAPE . MAXVALUE , FAMILY . UNIQUE . HIGH NOWAIT FOR UPDATE OF GREATEST , FAMILY . ident . FAMILY , INDEX_AFTER_ORDER_BY_BEFORE_AT , INDEX_AFTER_ORDER_BY_BEFORE_AT , ident . NAN , FAMILY SKIP LOCKED ) TRANSFORM FROM [ FUNCTION -247 ] ( error WITH ORDINALITY WHERE ROW ( ) OVERLAPS ( ) IS NULL IS NOT FALSE
    rsg_test.go:125: Last executed (37): SELECT UNIQUE ( error CONTAINS DEFAULT NOTNULL FROM ident 
    rsg_test.go:125: Last executed (38): SELECT * , * , ROW ( ) OVERLAPS ROW ( ) JOB FROM ident 
    rsg_test.go:125: Last executed (39): SELECT DEFAULT = ROW ( ) OVERLAPS ROW ( ) AS ident FROM ( NOTHING_AFTER_RETURNING . TRANSACTIONS @ { FAMILY = [ 452 ] } WITH ORDINALITY NATURAL JOIN ONLY ( INDEX_BEFORE_PAREN . COMMITTED . FAMILY ) @ { FORCE_INDEX = CURRENT_TIMESTAMP } ) WITH ORDINALITY ident ( ISERROR , ident ) WHERE ROW ( ) IS NOT NULL
    rsg_test.go:125: Last executed (40): SELECT ( ( TABLE error ) ORDER BY PRIMARY KEY INDEX_BEFORE_NAME_THEN_PAREN . METHOD . ident DESC , INDEX_AFTER_ORDER_BY_BEFORE_AT SUBSTRING . ON . OUT @ ident ASC ) AS NULL FROM [ WITH ident ( BOX2D ) AS MATERIALIZED ( TRUNCATE error ) INSERT error ] WHERE OPERATOR ( FORCE_INDEX . ~ ) DEFAULT
    rsg_test.go:125: Last executed (41): SELECT UNIQUE ( error COLLATE COLUMNS PRIVILEGES FROM CURRENT_TIMESTAMP AS ( ident POLYGON error ARRAY ) WHERE EXISTS ( WITH RECURSIVE ident AS ( IMPORT error ) ( error ) TYPEANNOTATE POLYGON error ARRAY [ -201 ] [ error
    rsg_test.go:125: Last executed (42): SELECT @ -13 AS FAMILY FROM FAMILY * @ INDEX WHERE ( ) OVERLAPS ( ) ILIKE EXISTS ( WITH GEOMETRYCOLLECTIONZM AS NOT MATERIALIZED ( RESET CLUSTER error ) ( error ORDER BY PRIMARY KEY ident . LEAST )
    rsg_test.go:125: Last executed (43): SELECT DEFAULT TYPECAST POINT error ARRAY [ -421 ] [ error AS EXTREMES , * FROM INDEX_BEFORE_NAME_THEN_PAREN . LOCALTIME @ { FORCE_INDEX = [ 461 ] , FAMILY = [ 232 ] } AS ident WHERE INTERVAL ( -101 ) 'string' IS UNKNOWN NOTNULL
    rsg_test.go:125: Last executed (44): SELECT * , DEFAULT AS NOCREATEROLE FROM [ EXPLAIN ANALYSE ( FAMILY , LATEST , OUTER ) COMMENT ON COLUMN INDEX_BEFORE_NAME_THEN_PAREN . ORDER IS 'string' ] TRIGGERS WHERE ROW ( ) OVERLAPS ROW ( ) ^ UNIQUE ( error
    rsg_test.go:125: Last executed (45): SELECT * , ROW ( ) OVERLAPS ( ) IN ( ( error FOR NO KEY UPDATE ) IS NULL FROM ident 
    rsg_test.go:125: Last executed (46): SELECT B'10010' IS NOT FALSE AS LATEST FROM [ UPSERT error ] WITH ORDINALITY AS HOUR WHERE DEFAULT IS NOT OF ( POLYGON error ARRAY [ 361 ] [ error )
    rsg_test.go:125: Last executed (47): SELECT ( ) OVERLAPS ( ) NOCREATEDB FROM ONLY FAMILY @ [ -332 ] WHERE CBRT EXISTS ( ( error ORDER BY INDEX_AFTER_ORDER_BY_BEFORE_AT FAMILY . FORCE . FAMILY @ ident )
    rsg_test.go:125: Last executed (48): SELECT * , DEFAULT IS NOT UNKNOWN AS ident , * FROM [ -281 ( 53 , -379 ) VIRTUAL ] @ [ -278 ] WITH ORDINALITY AS POINT ( YEAR , BOX2D ) WHERE EXISTS ( WITH BETWEEN AS MATERIALIZED ( BACKUP TABLE CSV . CONSTRAINT TO error ) , TIMESTAMP ( COMMITTED ) AS NOT MATERIALIZED ( UPSERT error ) , WORK AS ( EXPERIMENTAL SCRUB error ) , ident AS MATERIALIZED ( INSERT error ) ( error ORDER BY PRIMARY KEY FAMILY . NULLIF ) IS UNKNOWN
    rsg_test.go:125: Last executed (49): SELECT PLACEHOLDER IS OF ( @ 250 ARRAY , POINT error [ ] , VECTOR ( 426 ) [ ] , INTERVAL HOUR TO MINUTE ARRAY ) IS NULL AS RIGHT FROM [ TABLE error FETCH FIRST ROWS ONLY ] WITH ORDINALITY ident ( IMPORT , LOCKED , RETURN , ident , ident , INTEGER ) WHERE NOT EXISTS ( ( TABLE error ) )
    rsg_test.go:128: Currently active statements: map[]
    rsg_test.go:400: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/d437d2c847dfedbc4972f231c3331c8e/logTestRandomSyntaxSelect1587468282
--- FAIL: TestRandomSyntaxSelect (300.75s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxSelect.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-62924

--- a/pkg/crosscluster/logical/sqlwriter/testdata/ldr_statements
+++ b/pkg/crosscluster/logical/sqlwriter/testdata/ldr_statements
@@ -22,7 +22,7 @@ DELETE FROM [104 AS replication_target] WHERE (((id = $1::INT8) AND (name IS NOT
 
 show-select table=basic_table
 ----
-SELECT key_list.index, replication_target.crdb_internal_origin_timestamp, replication_target.crdb_internal_mvcc_timestamp, replication_target.id, replication_target.name, replication_target.value, replication_target.data FROM ROWS FROM (unnest($1::INT8[])) WITH ORDINALITY AS key_list (key1, index) INNER LOOKUP JOIN [104 AS replication_target] ON replication_target.id = key_list.key1
+SELECT key_list.index, replication_target.crdb_internal_origin_timestamp, replication_target.crdb_internal_mvcc_timestamp, replication_target.id, replication_target.name, replication_target.value, replication_target.data FROM ROWS FROM (unnest($1::INT8[])) WITH ORDINALITY AS key_list (key1, "index") INNER LOOKUP JOIN [104 AS replication_target] ON replication_target.id = key_list.key1
 
 exec
 CREATE TYPE status AS ENUM ('pending', 'active', 'completed', 'cancelled');
@@ -56,7 +56,7 @@ DELETE FROM [107 AS replication_target] WHERE (((((id = $1::INT8) AND (title IS 
 
 show-select table=tasks
 ----
-SELECT key_list.index, replication_target.crdb_internal_origin_timestamp, replication_target.crdb_internal_mvcc_timestamp, replication_target.id, replication_target.title, replication_target.description, replication_target.status, replication_target.priority, replication_target.created_at FROM ROWS FROM (unnest($1::INT8[])) WITH ORDINALITY AS key_list (key1, index) INNER LOOKUP JOIN [107 AS replication_target] ON replication_target.id = key_list.key1
+SELECT key_list.index, replication_target.crdb_internal_origin_timestamp, replication_target.crdb_internal_mvcc_timestamp, replication_target.id, replication_target.title, replication_target.description, replication_target.status, replication_target.priority, replication_target.created_at FROM ROWS FROM (unnest($1::INT8[])) WITH ORDINALITY AS key_list (key1, "index") INNER LOOKUP JOIN [107 AS replication_target] ON replication_target.id = key_list.key1
 
 # Test a table with a stored and virtual computed column. Note that the
 # computed columns are not included in the INSERT/UPDATE/DELETE statements.
@@ -92,7 +92,7 @@ DELETE FROM [108 AS replication_target] WHERE ((((id = $1::INT8) AND (name IS NO
 # discount_price is included because its part of the primary key.
 show-select table=products
 ----
-SELECT key_list.index, replication_target.crdb_internal_origin_timestamp, replication_target.crdb_internal_mvcc_timestamp, replication_target.id, replication_target.name, replication_target.unit_price, replication_target.quantity, replication_target.total_value, replication_target.last_updated FROM ROWS FROM (unnest($1::INT8[], $2::DECIMAL[])) WITH ORDINALITY AS key_list (key1, key2, index) INNER LOOKUP JOIN [108 AS replication_target] ON (replication_target.id = key_list.key1) AND (replication_target.total_value = key_list.key2)
+SELECT key_list.index, replication_target.crdb_internal_origin_timestamp, replication_target.crdb_internal_mvcc_timestamp, replication_target.id, replication_target.name, replication_target.unit_price, replication_target.quantity, replication_target.total_value, replication_target.last_updated FROM ROWS FROM (unnest($1::INT8[], $2::DECIMAL[])) WITH ORDINALITY AS key_list (key1, key2, "index") INNER LOOKUP JOIN [108 AS replication_target] ON (replication_target.id = key_list.key1) AND (replication_target.total_value = key_list.key2)
 
 # Test a table with an expression, inverted index, and partial index. The
 # indexes are not expected to impact the INSERT/UPDATE/DELETE statements.
@@ -141,7 +141,7 @@ DELETE FROM [109 AS replication_target] WHERE ((((((id = $1::INT8) AND (first_na
 
 show-select table=employees
 ----
-SELECT key_list.index, replication_target.crdb_internal_origin_timestamp, replication_target.crdb_internal_mvcc_timestamp, replication_target.id, replication_target.first_name, replication_target.last_name, replication_target.email, replication_target.salary, replication_target.department, replication_target.hire_date FROM ROWS FROM (unnest($1::INT8[])) WITH ORDINALITY AS key_list (key1, index) INNER LOOKUP JOIN [109 AS replication_target] ON replication_target.id = key_list.key1
+SELECT key_list.index, replication_target.crdb_internal_origin_timestamp, replication_target.crdb_internal_mvcc_timestamp, replication_target.id, replication_target.first_name, replication_target.last_name, replication_target.email, replication_target.salary, replication_target.department, replication_target.hire_date FROM ROWS FROM (unnest($1::INT8[])) WITH ORDINALITY AS key_list (key1, "index") INNER LOOKUP JOIN [109 AS replication_target] ON replication_target.id = key_list.key1
 
 exec
 ALTER DATABASE defaultdb PRIMARY REGION "us-east1";
@@ -175,7 +175,7 @@ DELETE FROM [112 AS replication_target] WHERE (((((id = $1::UUID) AND (user_id I
 
 show-select table=user_events
 ----
-SELECT key_list.index, replication_target.crdb_internal_origin_timestamp, replication_target.crdb_internal_mvcc_timestamp, replication_target.id, replication_target.user_id, replication_target.event_type, replication_target.event_data, replication_target.created_at, replication_target.region FROM ROWS FROM (unnest($1::UUID[], $2::@100111)) WITH ORDINALITY AS key_list (key1, key2, index) INNER LOOKUP JOIN [112 AS replication_target] ON (replication_target.id = key_list.key1) AND (replication_target.region = key_list.key2)
+SELECT key_list.index, replication_target.crdb_internal_origin_timestamp, replication_target.crdb_internal_mvcc_timestamp, replication_target.id, replication_target.user_id, replication_target.event_type, replication_target.event_data, replication_target.created_at, replication_target.region FROM ROWS FROM (unnest($1::UUID[], $2::@100111)) WITH ORDINALITY AS key_list (key1, key2, "index") INNER LOOKUP JOIN [112 AS replication_target] ON (replication_target.id = key_list.key1) AND (replication_target.region = key_list.key2)
 
 # Test table with array primary key - this requires point select due to array-of-arrays limitations
 exec

--- a/pkg/sql/parser/testdata/select_clauses
+++ b/pkg/sql/parser/testdata/select_clauses
@@ -2877,6 +2877,59 @@ SELECT (*) FROM ROWS FROM ((json_to_record(('')))) AS t (a "Nice Enum 📙", b S
 SELECT * FROM ROWS FROM (json_to_record('_')) AS t (a "Nice Enum 📙", b STRING, c foo) -- literals removed
 SELECT * FROM ROWS FROM (_('')) AS _ (_ _, _ STRING, _ _) -- identifiers removed
 
+# Regression test for #168372: a column name in an AS column-def list must be
+# quoted when it is the unreserved keyword INDEX, because the lexer's INDEX
+# disambiguation heuristic reclassifies a bare INDEX followed by an identifier
+# and '(' (e.g. a parameterized type like VARBIT(61)) into
+# INDEX_BEFORE_NAME_THEN_PAREN, which the 'name' grammar production rejects.
+#
+# The input uses BIT VARYING(61) (a two-word type) so the initial parse
+# succeeds: the lexer only sees INDEX followed by BIT followed by VARYING (not
+# '('). The formatter normalizes the type to VARBIT(61), which would trigger
+# the reclassification on re-parse if INDEX were left bare.
+
+parse
+SELECT * FROM json_to_record('') AS t (index BIT VARYING(61))
+----
+SELECT * FROM ROWS FROM (json_to_record('')) AS t ("index" VARBIT(61)) -- normalized!
+SELECT (*) FROM ROWS FROM ((json_to_record(('')))) AS t ("index" VARBIT(61)) -- fully parenthesized
+SELECT * FROM ROWS FROM (json_to_record('_')) AS t ("index" VARBIT(61)) -- literals removed
+SELECT * FROM ROWS FROM (_('')) AS _ (_ VARBIT(61)) -- identifiers removed
+
+parse
+SELECT * FROM json_to_record('') AS t (index BIT VARYING(61)[])
+----
+SELECT * FROM ROWS FROM (json_to_record('')) AS t ("index" VARBIT(61)[]) -- normalized!
+SELECT (*) FROM ROWS FROM ((json_to_record(('')))) AS t ("index" VARBIT(61)[]) -- fully parenthesized
+SELECT * FROM ROWS FROM (json_to_record('_')) AS t ("index" VARBIT(61)[]) -- literals removed
+SELECT * FROM ROWS FROM (_('')) AS _ (_ VARBIT(61)[]) -- identifiers removed
+
+parse
+SELECT * FROM json_to_record('') AS t (a INT, index BIT VARYING(61), b INT)
+----
+SELECT * FROM ROWS FROM (json_to_record('')) AS t (a INT8, "index" VARBIT(61), b INT8) -- normalized!
+SELECT (*) FROM ROWS FROM ((json_to_record(('')))) AS t (a INT8, "index" VARBIT(61), b INT8) -- fully parenthesized
+SELECT * FROM ROWS FROM (json_to_record('_')) AS t (a INT8, "index" VARBIT(61), b INT8) -- literals removed
+SELECT * FROM ROWS FROM (_('')) AS _ (_ INT8, _ VARBIT(61), _ INT8) -- identifiers removed
+
+# Already-quoted INDEX as a column name continues to round-trip.
+parse
+SELECT * FROM json_to_record('') AS t ("index" BIT VARYING(61))
+----
+SELECT * FROM ROWS FROM (json_to_record('')) AS t ("index" VARBIT(61)) -- normalized!
+SELECT (*) FROM ROWS FROM ((json_to_record(('')))) AS t ("index" VARBIT(61)) -- fully parenthesized
+SELECT * FROM ROWS FROM (json_to_record('_')) AS t ("index" VARBIT(61)) -- literals removed
+SELECT * FROM ROWS FROM (_('')) AS _ (_ VARBIT(61)) -- identifiers removed
+
+# Reduced reproducer of the original RandomSyntaxSelect failure (#168372).
+parse
+SELECT * FROM [FUNCTION 198]() AS TEMP (INDEX BIT VARYING(61) ARRAY[233])
+----
+SELECT * FROM ROWS FROM ([FUNCTION 198]()) AS temp ("index" VARBIT(61)[]) -- normalized!
+SELECT (*) FROM ROWS FROM (([FUNCTION 198]())) AS temp ("index" VARBIT(61)[]) -- fully parenthesized
+SELECT * FROM ROWS FROM ([FUNCTION 198]()) AS temp ("index" VARBIT(61)[]) -- literals removed
+SELECT * FROM ROWS FROM ([FUNCTION 198]()) AS _ (_ VARBIT(61)[]) -- identifiers removed
+
 parse
 SELECT 123 AS OF FROM t
 ----

--- a/pkg/sql/sem/tree/select.go
+++ b/pkg/sql/sem/tree/select.go
@@ -18,6 +18,7 @@ import (
 	"fmt"
 	"slices"
 
+	"github.com/cockroachdb/cockroach/pkg/sql/lexbase"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/catid"
@@ -213,11 +214,25 @@ type ColumnDef struct {
 
 // Format implements the NodeFormatter interface.
 func (c *ColumnDef) Format(ctx *FmtCtx) {
-	ctx.FormatNode(&c.Name)
+	if !ctx.flags.HasFlags(FmtAnonymize) && needsLexerEscapeAsColName(string(c.Name)) {
+		lexbase.EncodeEscapedSQLIdent(&ctx.Buffer, string(c.Name))
+	} else {
+		ctx.FormatNode(&c.Name)
+	}
 	if c.Type != nil {
 		ctx.WriteByte(' ')
 		ctx.FormatTypeReference(c.Type)
 	}
+}
+
+// needsLexerEscapeAsColName returns true when n must be force-quoted to be
+// recognized as a column name by the parser, even though it is not a reserved
+// keyword. This is necessary for identifiers that the lexer reclassifies into
+// a non-name token based on lookahead (e.g. INDEX followed by an identifier
+// and '(' becomes INDEX_BEFORE_NAME_THEN_PAREN). See the INDEX disambiguation
+// comment in pkg/sql/parser/lexer.go for details.
+func needsLexerEscapeAsColName(n string) bool {
+	return n == "index"
 }
 
 // ColumnDefList represents a list of ColumnDefs.

